### ISSUE TITLE
Export with embedded pck is broken when changing PE metadata with rcedit

### ISSUE THREAD

**Issue by Wavesonics** [godotengine/godot#33466]
**Godot version:** 3.2 beta 1

**OS/device including version:**
Windows 10

**Issue description:**
Running an executable that was exported with embedded pck fails to run with the following error:
`Error: Could not load game data at path '.'. Is the .pck file missing?`

**Steps to reproduce:**

1. Enable "Embeded Pck" in export options
2. Export (_either release or debug, doesn't matter_)

**Minimal reproduction project:**

**Comment by RandomShaper** [comment]
@akien-mga, I'll check this more thoroughly later, but for now it'd be helpful knowing if the export templates are signed.

**Comment by akien-mga** [comment]
They are signed indeed, but as per #32310 this shouldn't prevent the embedded PCK from working, it should only lose the signature on export.

But maybe @Wavesonics signed the exported binary themselves, in which case it would be a duplicate of #32310.

**Comment by Wavesonics** [comment]
@akien-mga I did not sign my windows binaries, nor is Code signing is not enabled in my export options.

**Comment by Wavesonics** [comment]
I went to make a minimal case project for this bug, but may have instead uncovered a different bug: #33549

**Comment by Wavesonics** [comment]
Ah I think I found the problem. I am using `rcedit-x64.exe` to change the icon for the Windows exe, that must screw up the signing that @akien-mga is talking about?

In my windows export options I have an icon specified under the **Application** header.

If I remove that, it runs just fine, presumable because it doesn't run `rcedit-x64.exe`?

**Comment by Wavesonics** [comment]
If the issue is the fact that the template is signed, and then this icon program modifies the template, thus invalidating the signature. Are we in "how did this ever work" territory?

**Comment by bruvzg** [comment]
- If export template is signed (why is it?), both using `rcedit` and embedding PCK will break signature and result in corrupted executable, that can't be signed later. But neither shouldn't prevent it from running.
- If executable is signed after embedding PCK, it won't run.
- If `rcedit` changes are applied after embedding PCK, it won't run (this is most likely the case here).

**Comment by akien-mga** [comment]
At the very least we should add clear documentation on the export tutorials about limitations of the embedded PCK.

Users are lured in by the "all in one" aspect but there are lots of non-obvious drawbacks to the approach.

**Comment by RandomShaper** [comment]
Yes, current problems with resource editing and signing could be considered known issues for the time being.
Both can be fixed. I'll work on it the next chance I have.

**Comment by akien-mga** [comment]
>     * If export template is signed (why is it?), both using `rcedit` and embedding PCK will break signature and result in corrupted executable, that can't be signed later. But neither shouldn't prevent it from running.

I disabled signing of export templates as of 3.2 RC 2.

**Comment by der-ac** [comment]
Is there any workaround for using PCK embedding and having a custom icon at the same time?

**Comment by RandomShaper** [comment]
There's a potential megahack: tampering the current icon image data in the file with the one you want to use.

**Comment by akien-mga** [comment]
You can replace the icon with rcedit in the export template itself, so that it's already good before exporting with embedded PCK.

**Comment by RandomShaper** [comment]
> You can replace the icon with rcedit in the export template itself, so that it's already good before exporting with embedded PCK.

Much better. :)

**Comment by Tsar333** [comment]
> You can replace the icon with rcedit in the export template itself, so that it's already good before exporting with embedded PCK.

Not quite sure how you do that. Are there instructions somewhere on this strategy ? many thanks. It is true that for non-professionnal gamedevs (speaking for myself), being able to ship one single .exe file does sound more appealing.

**Comment by Calinou** [comment]
[rcedit](https://github.com/electron/rcedit) is a command-line utility you can call manually. [Download an executable](https://github.com/electron/rcedit/releases) and run it from a command-line for instructions.

**Comment by eyeEmotion** [comment]
So... what are the constraints now? Because one time I can export an exe with the pck embed in it, but most of the time not. That while I didn't change my export settings. And I've done all the .ico and .png changes.

**Comment by eyeEmotion** [comment]
Ok, this is weird. Although my game works fine in Godot (F5 or F6), when exporting it, there is all sort of weird things happening. Previous mentioned problem, but also game breaking problem.
I have some code in my game to detect if the input is from a keyboard or controller. If it's from the keyboard, certain camera-effects shouldn't be executed. This works fine in Godot itself... and when finally able to export the game, it also does work with the executable... but only for a while.

When I just exported it and tested it, everything worked fine. Now, a few hours later testing it again, the problem occurs again. And I also noticed that the icon image of my executable has changed from the one I made/chose, to the one of Godot again. So what is happening here? Why is the executable fine at first and a few hours later, while doing nothing with it, this happens.
Hope this gets fixed, because if you can't export your game, why bother then.

**Comment by Calinou** [comment]
@eyeEmotion If you export your project with a non-embedded PCK then export it again with an embedded PCK, the non-embedded PCK will take priority. This will be fixed by #38308.

As a workaround, remove the PCK file that's next to the exported project executable.

**Comment by eyeEmotion** [comment]
I've done all those things. I've always first removed the old files before exporting the new files.
Removing the "personal" icon in the export seems to help to be able to export the embed pck. But whatever I do, after a while my the game "breaks" some code in trying to execute a function it shouldn't be able to execute. This only happens with the exported file. Don't know if it happens with the Linux/MAC OS version, since I don't have them (although they are available to download on my itch.io game-page.
And it's just a simple code: detect if _input is InputEventKey, set variable either to true or false. That variable is used in another function that will only execute it when it's a controller. So pretty straightforward.
It's just so strange that the first few starts of the executable, this works just fine and then all of a sudden, it doesn't anymore. Even with the seperate exe and pck.

**Comment by follower** [comment]
I investigated the use of `rcedit` recently and encountered [this comment](https://github.com/electron/rcedit/issues/91#issuecomment-613852694) which suggests it may no longer be a reliable option.

(I did note that [Kaitai Struct does have a PE specification file](http://formats.kaitai.io/microsoft_pe/index.html) if anyone wants to code an alternative. ;D )

**Comment by kukosek** [comment]
funny bug, yesterday we were exporting our game to windows (last day of game off game jam) used rcedit only to find out we cant use embedded pck then.

**Comment by RandomShaper** [comment]
Just as a reminder, I know there are multiple issues with PCK embedding. I have plans to provide a better implementation, that is not as fragile as the current one. I don't know exactly when I'll be able to do it, but that'll happen.

**Comment by MikeTeok** [comment]
Holy sh**ttt, I found out that as long as I filled File Version and Product Version, this issue will be gone.

**Comment by madelfagam** [comment]
> Holy sh**ttt, I found out that as long as I filled File Version and Product Version, this issue will be gone.

yes it really works! I indicated random numbers in these two fields and after export everything worked out.

**Comment by madelfagam** [comment]
> yes it really works! I indicated random numbers in these two fields and after export everything worked out.

Oops, it seems this only works if the executable is at the root folder of the game's resources ...

**Comment by MikeTeok** [comment]
> > yes it really works! I indicated random numbers in these two fields and after export everything worked out.
> 
> Oops, it seems this only works if the executable is at the root folder of the game's resources ...

Weird, I didn't export at the game's folder, I exported far far away from source code. I filled with 1.0.0 instead of random number tho. Not sure if it matters.

**Comment by nathanfranke** [comment]
I can reproduce this on Linux using `strip`.

```
[nathan@arch .export]$ ./Test.x86_64 
Godot Engine v3.3.4.stable.official.faf3f883d - https://godotengine.org
OpenGL ES 3.0 Renderer: NVIDIA GeForce GTX 960/PCIe/SSE2
OpenGL ES Batching: ON
 
[nathan@arch .export]$ strip Test.x86_64 -o test

[nathan@arch .export]$ ./test
Error: Couldn't load project data at path ".". Is the .pck file missing?
If you've renamed the executable, the associated .pck file should also be renamed to match the executable's name (without the extension).
```

GNU strip (GNU Binutils) 2.36.1
Kernel 5.10.76-1-lts

Project is exported on 3.3.4:
[Test9_1.zip](https://github.com/godotengine/godot/files/7466654/Test9_1.zip)

**Comment by Deozaan** [comment]
I was pretty sure this got fixed sometime in 3.3.x and I didn't have to resort to weird hacks to get a custom icon and metadata while embedding the PCK file. Maybe I was wrong about that.

But I'm definitely experiencing it again in 3.4.

**Comment by magni-inter-pix** [comment]
Yeah, I’m in the same situatie. I’m close to finalizing my project and it randomly happens when exporting to windows.
I just tested with 3.4.1 RC1 and that bug is still there
Sometimes I get an error while exporting:
```
ERROR: Failed to export the project for platform 'Windows Desktop'.
   at: _export_project_to_path (editor/project_export.cpp:863)
```

Sometimes it does export, but I get an error when starting the exe:
```
Error: Couldn't load project data at path ".". Is the .pck file missing?
If you've renamed the executable, the associated .pck file should also be renamed to match the executable's name (without the extension)
```
Sometimes it did export good  from 3.4.0, randomly
 
When I export with the checkbox “Embed pck” unchecked, the export does work 
The problem is because of rcedit as mentioned here:
[https://github.com/godotengine/godot/issues/45345#issuecomment-798772388 ](https://github.com/godotengine/godot/issues/45345#issuecomment-798772388 )
It started after I added icons. Disabling rcedit also fixes the issue, but I want the exported executable in one .exe with an icon, so I’m going to need to use resource hacker to get my windows icon in when this is not fixed shortly I presume.

**Comment by Eric-pixel-0** [comment]
My project has worked normally
I can't find what I do

https://github.com/Eric-pixel-0/memoriaMusical.git

**Comment by Calinou** [comment]
Can anyone check if https://github.com/godotengine/godot/pull/57850 fixes this issue? You can try 4.0alpha5 or building the `3.x` branch from source (both editor *and* Windows export template)?

**Comment by nathanfranke** [comment]
I can still reproduce on 11d40d76a with my steps above.

```
$ ./test.x86_64 
Godot Engine v3.5.beta.custom_build.11d40d76a - https://godotengine.org
OpenGL ES 3.0 Renderer: NVIDIA GeForce GTX 960/PCIe/SSE2
Async. shader compilation: OFF
 
$ strip test.x86_64 
$ ./test.x86_64 
Error: Couldn't load project data at path ".". Is the .pck file missing?
If you've renamed the executable, the associated .pck file should also be renamed to match the executable's name (without the extension).
```

I think my issue is not caused by the windows icon, but an underlying way that the pck file is embedded.
